### PR TITLE
Updates informer pkg to use TransformFunc()

### LINKED ISSUE #24923: Make use of the TransFunc introduced in k8s 1.27
Kubernetes has introduced in https://github.com/kubernetes/kubernetes/pull/116623, and backported to all branches, the TransFunc that allows to mutate certain resources in K8s Caches. This will allow us to remove our own TransFunc implementation functions in https://github.com/cilium/cilium/blob/264dd3f350022c7626438f2dab0401dcb41ec616/pkg/k8s/informer/informer.go#L101-L106 by using the DeltaFIFO directly from the library.

Depending how non-invasive the patch will be, we might consider backporting it to our stable branches as well.

cc @odinuge

### PR DESCRIPTION
**Pull request by danehans**
- Removes `ConvertFunc` in favor of [TransformFunc](https://pkg.go.dev/k8s.io/client-go@v0.26.0/tools/cache#TransformFunc) for object transformation. Informers are now created with `TransformFunc` instead of `ConvertFunc` to mutate certain resources in K8s Caches.
- Renames ConvertXXX functions to TransformXXX.
- Adds tests to `pkg/k8s/init_test.go` to improve test coverage of transformer functions.

Fixes: #24923

### CODE REVIEW

File: operator/watchers/cilium_endpoint.go
Comment (christarazi) [inline operator/watchers/cilium_endpoint.go]:
```suggestion
			return nil, fmt.Errorf("unknown object type %T", concreteObj.Obj)
```

Please apply this to all other instances

File: operator/cmd/root.go
Comment (christarazi) [inline operator/cmd/root.go]: There should be some relevant logfields in `pkg/logging/logfields` that we can use here without making the log msg dynamic. This enables the use of structured logging.

```suggestion
								scopedLog.WithError(err).WithFields({
								  ...
								}.)Error("Failed to transform k8s service")
```

File: operator/cmd/root.go
Comment (christarazi) [inline operator/cmd/root.go:532]: Same here regarding structured logging

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR! A few comments to address below.

File: pkg/k8s/informer/informer.go
Comment (giorio94) [inline pkg/k8s/informer/informer.go]: Given that this function is now a vanilla copy of the upstream one, I would suggest to update the callers to use the upstream one instead (either `NewInformer` or `NewTransformingInformer` depending on whether the transformer function is nil).

Comment (joamaki) [reply pkg/k8s/informer/informer.go]: Looks like `pkg/k8s/informer` does two things: it allows using a custom store and it injects a `CacheMutationDetector`. The custom store we can't do directly with the upstream informer, but the latter can be done with a transforming informer. Since we can't get rid of the custom store use yet, it's probably easier to leave the callers as is for now.

EDIT: Ah, I don't think we actually enable the mutation detection anywhere (`git grep KUBE_CACHE_MUTATION_DETECTOR` )

Comment (giorio94) [reply pkg/k8s/informer/informer.go]: Yeah, I was referring to `NewTransformingInformer` only, which doesn't take a custom store as parameter. But I missed that there is currently the custom `CacheMutatorDetector` part. Given that, I agree that it is probably better to not modify the callers for the moment (I would suggest to update the comment though, since it is currently a bit misleading).

File: pkg/nodediscovery/nodediscovery.go
Comment (danehans) [inline pkg/nodediscovery/nodediscovery.go:451]: __Note:__ This transformation was removed since https://github.com/cilium/cilium/pull/25282 updated `GetK8sNode()` to return a slim node instead of a k8s node.

File: pkg/k8s/informer/informer.go
Comment (giorio94) [inline pkg/k8s/informer/informer.go]: Nit. The MutationDetector is actually unrelated from the Store.

File: pkg/k8s/informer/informer.go
Comment (giorio94) [inline pkg/k8s/informer/informer.go]: Nit. I wonder if it could be worth to revert the renaming of this function (and of the one below) to reduce the code churn. I don't have a strong opinion though.

**Review by giorio94 — APPROVED** [review]
Thanks! Changes LGTM

Just a couple of nits inline. Could you please also extend the commit message with the context around this change?

**Comment by danehans** [comment]
@giorio94 thanks for the /approval. I updated the PR description and commit 44598e0 resolves your nits.

**Comment by giorio94** [comment]
> @giorio94 thanks for the /approval. I updated the PR description and commit [44598e0](https://github.com/cilium/cilium/commit/44598e0a48fd040eeafa5871a14d58fd58130681) resolves your nits.

Thanks! Could you please also extend the commit message with a bit of description?

**Review by mhofstetter — COMMENTED** [review]
LGTM, thanks!

**Review by mhofstetter — APPROVED** [review]

File: pkg/nodediscovery/nodediscovery.go
Comment (gandro) [inline pkg/nodediscovery/nodediscovery.go:449]: This seems to have been the last use of `k8s.ConvertToNode` function outside of test code. We should probably remove it (doesn't have to be done in this PR though)

**Review by gandro — APPROVED** [review]
Thanks!

File: pkg/nodediscovery/nodediscovery.go
Comment (danehans) [reply pkg/nodediscovery/nodediscovery.go:449]: @gandro thanks for catching this. I created https://github.com/cilium/cilium/issues/25680 to track the removal of the func.

**Comment by danehans** [comment]
@christarazi you still have requested changes that I resolved from your review. PTAL when you have a moment.

**Review by christarazi — APPROVED** [review]

**Review by joamaki — APPROVED** [review]

File: pkg/k8s/informer/informer.go
Comment (joamaki) [inline pkg/k8s/informer/informer.go:101]: What's the behavior of the informer when we return an error here?

Comment (danehans) [reply pkg/k8s/informer/informer.go:101]: The informer will not process the object. For example, the clustermesh-apiserver creates an informer that transforms k8s nodes to cilium nodes. If `TransformToCiliumNode()` returns an error, the provided object is not created/updated. We could set `RetryOnError` to re-enqueue the object. Thoughts?

**Comment by danehans** [comment]
@christarazi and others, is there anything else needed to merge this PR?

**Comment by danehans** [comment]
Commit c2078e9 rebases due to https://github.com/cilium/cilium/pull/25049.

**Comment by danehans** [comment]
@christarazi thanks for kicking off the tests. The `test-1.26-net-next` job is failing due to:

```sh
/home/jenkins/workspace/Cilium-PR-K8s-1.26-kernel-net-next/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:515
Unexpected missed tail call
Expected
    <int>: 1
to be ==
    <int>: 0
/home/jenkins/workspace/Cilium-PR-K8s-1.26-kernel-net-next/src/github.com/cilium/cilium/test/k8s/updates.go:129
```

It appears the upgrade was unsuccessful (^ calls `assertUpgradeSuccessful()`), possibly due to one of the agents failing readiness probes:
```sh
22:18:19 STEP: Cilium is not ready yet: connectivity health is failing: Cluster connectivity is unhealthy on 'cilium-rwcn6': Exitcode: 1 
Err: exit status 1
Stdout:
 	 
Stderr:
 	 Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
	 Error: Cannot get status/probe: Put "[http://%2Fvar%2Frun%2Fcilium%2Fhealth.sock/v1beta/status/probe":](http://%2Fvar%2Frun%2Fcilium%2Fhealth.sock/v1beta/status/probe) dial unix /var/run/cilium/health.sock: connect: no such file or directory
	 
	 command terminated with exit code 1
```

I don't believe this is related to the PR so I wanted to get your input on how to proceed.

**Comment by christarazi** [comment]
It's flake https://github.com/cilium/cilium/issues/24514.

**Comment by christarazi** [comment]
Let's see if the next test run can pass without flakes.

**Comment by danehans** [comment]
@christarazi when you have a moment, can you please advise me on how to move this PR forward?

**Comment by joestringer** [comment]
@danehans once the reviews are all approved from relevant code owner areas and the required CI tests are all passing, the maintainer's little helper marks the PR ready-to-merge like above. Someone then comes through and does last sanity checks and merges the PR. Thanks for the contribution!

**Comment by odinuge** [comment]
cc @joestringer @danehans @christarazi ^

**Comment by odinuge** [comment]
Also, feel free to ping me on PRs around this part of client-go, given I re(wrote) some of it a few months ago to fix some policy leaks we saw in cilium 😄

**Comment by danehans** [comment]
@odinuge thanks for the feedback. As you can see [here](https://github.com/cilium/cilium/pull/25604/files#diff-fec30a88ffb515619228eb30191a763c3df4d6131791e74a6a2fae2b4265b69dL558-L599), this PR did not change the logic of `TransformToCCNP()`. Prior to this PR, the code you're referencing looked like:
```go
case cache.DeletedFinalStateUnknown:
		ccnp, ok := concreteObj.Obj.(*cilium_v2.CiliumClusterwideNetworkPolicy)
		if !ok {
			return obj
		}
```

and now the same codeblock returns nil and an error instead of the provided object:
```go
	case cache.DeletedFinalStateUnknown:
		ccnp, ok := concreteObj.Obj.(*cilium_v2.CiliumClusterwideNetworkPolicy)
		if !ok {
			return nil, fmt.Errorf("unknown object type %T", concreteObj.Obj)
		}
```

`TransformToCCNP()` before and after this PR expected `obj` to be a `CiliumClusterwideNetworkPolicy` (embedded in the case of `DeletedFinalStateUnknown`) and return a `SlimCNP` (embedded in `DeletedFinalStateUnknown`). This pattern is consistent with other TransformXXX funcs in the pkg. The associated tests were updated accordingly and currently pass when I run them locally. It appears that `TransformToCCNP()` is being called somewhere in the code when it shouldn't be, e.g. no need to transform an object that is already a `SlimXXX`.

**Comment by odinuge** [comment]
Thanks for looking into this @danehans!

Yeah, the old godocs were pretty ambiguous, and the k8s behavior in `client-go` that the `informer` mimics, was pretty inconsistent previously. It still has some legacy it has to deal with, and i think thats the reason for the current function . I have also spent a lot of time wrapping my head around this. So I very much agree this is a bit confusing 😅 

I think it all boils down to this; https://github.com/cilium/cilium/blob/a82f0fc303c5ff511b2550b775849abec30dbb3e/pkg/k8s/informer/informer.go#L99-L102

There, we always call the transform funcs, just like `client-go`. So even if the object is an `DeletedFinalStateUnknown`, we run the transformer.

I should also have been explicit that my comment applies equally to all those transformer functions that return a different type than they receive; not only the CCNP one. 😄 

Should we maybe open a new issue to track this? What do you think @danehans? If you to implement a fix here, I'm happy to review! 😄  (tho. I'm heading out of the office for the next two weeks 🌴)

**Comment by danehans** [comment]
I assigned myself https://github.com/cilium/cilium/issues/26240 and will push a PR to fix it.

**Pull request merged**
